### PR TITLE
Sandbox friendly build

### PR DESCRIPTION
**Pull request by npatsakula**
Hi, @tazz4843 !

I've made some effort to simplify cmake machinery and make build sandbox-friendly (now it is not changing source directory), now this crate can be build with NIX Flake.

Any feedback appreciable!

### CODE REVIEW

File: sys/build.rs [hunk 1]
```diff
@@ -53,17 +51,26 @@ fn main() {
...
     println!("cargo:rerun-if-changed=wrapper.h");
 
+    let out = PathBuf::from(env::var("OUT_DIR").unwrap());
+    let whisper_root = out.join("whisper.cpp/");
+
+    if !whisper_root.exists() {
+        std::fs::create_dir_all(&whisper_root).unwrap();
+        fs_extra::dir::copy("./whisper.cpp", &out, &Default::default()).unwrap_or_else(|e| {
+            panic!(
+                "Failed to copy whisper sources into {}: {e}",
```
Comment (tazz4843) [inline sys/build.rs:61]: Minor nit: the inline format arg syntax is less readable here imo, the clippy lint should be ignored at the crate level.

File: sys/build.rs [hunk 2]
```diff
@@ -91,86 +95,34 @@ fn main() {
...
+        .define("WHISPER_ALL_WARNINGS", "OFF")
+        .define("WHISPER_ALL_WARNINGS_3RD_PARTY", "OFF")
+        .define("WHISPER_BUILD_TESTS", "OFF")
+        .define("WHISPER_BUILD_EXAMPLES", "OFF")
+        .pic(true);
+
+    if cfg!(feature = "coreml") {
+        config.define("WHISPER_COREML", "ON");
+        config.define("WHISPER_COREML_ALLOW_FALLBACK", "1");
+    } else if cfg!(feature = "cuda") {
```
Comment (tazz4843) [inline sys/build.rs:112]: I don't believe these features are mutually exclusive of each other, these shouldn't be else-if trees. Feel free to correct me if I'm wrong here.

**Review by tazz4843 — CHANGES_REQUESTED** [review]
Currently late in the day where I am, will test and possibly merge in the morning.

File: sys/build.rs (hunk 1 above)
Comment (npatsakula) [reply sys/build.rs]: Okey-dokey, fixed.

File: sys/build.rs (hunk 2 above)
Comment (npatsakula) [reply sys/build.rs]: Yes, you're right (at least it compiles with `openblas` and `coreml` features enabled). Fixed in last commit revision.

**Comment by tazz4843** [comment]
I don't run Windows so I can't help much with the failing check unfortunately.

**Review by tazz4843 — APPROVED** [review]
Besides the failing checks looks good to me

**Comment by npatsakula** [comment]
> I don't run Windows so I can't help much with the failing check unfortunately.

Don't worry, I found symbol search path at `whisper.cpp` CI. Looks like it should work now (last time only `rustfmt` was broken).

**Comment by npatsakula** [comment]
Hello @tazz4843! Can I ask you to run CI pipeline?

**Comment by tazz4843** [comment]
GitHub doesn't appear to be letting me run the checks this time around, perhaps force push once more?

**Comment by tazz4843** [comment]
Can't get this to compile locally:

```
memory 19GiB/62GiB at 10-15 09:01:26 exit 101 took 6s156ms
on git sandbox_friendly is pkg v0.9.0-rc.2 via rs v1.75.0-nightly
niko@niko-desktop whisper-rs > cargo run --example basic_use
   Compiling whisper-rs-sys v0.7.3 (/tmp/whisper-rs/sys)
The following warnings were emitted during compilation:

warning: Unable to generate bindings: clang diagnosed error: wrapper.h:1:10: fatal error: 'whisper.h' file not found
warning: Using bundled bindings.rs, which may be out of date

error: failed to run custom build command for `whisper-rs-sys v0.7.3 (/tmp/whisper-rs/sys)`

Caused by:
  process didn't exit successfully: `/tmp/whisper-rs/target/debug/build/whisper-rs-sys-c3f9ea4e385de8cf/build-script-build` (exit status: 101)
  --- stdout
  cargo:rustc-link-lib=dylib=stdc++
  cargo:rerun-if-changed=wrapper.h
  cargo:rerun-if-env-changed=TARGET
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS
  cargo:warning=Unable to generate bindings: clang diagnosed error: wrapper.h:1:10: fatal error: 'whisper.h' file not found

  cargo:warning=Using bundled bindings.rs, which may be out of date
  CMAKE_TOOLCHAIN_FILE_x86_64-unknown-linux-gnu = None
  CMAKE_TOOLCHAIN_FILE_x86_64_unknown_linux_gnu = None
  HOST_CMAKE_TOOLCHAIN_FILE = None
  CMAKE_TOOLCHAIN_FILE = None
  CMAKE_GENERATOR_x86_64-unknown-linux-gnu = None
  CMAKE_GENERATOR_x86_64_unknown_linux_gnu = None
  HOST_CMAKE_GENERATOR = None
  CMAKE_GENERATOR = None
  CMAKE_PREFIX_PATH_x86_64-unknown-linux-gnu = None
  CMAKE_PREFIX_PATH_x86_64_unknown_linux_gnu = None
  HOST_CMAKE_PREFIX_PATH = None
  CMAKE_PREFIX_PATH = None
  CMAKE_x86_64-unknown-linux-gnu = None
  CMAKE_x86_64_unknown_linux_gnu = None
  HOST_CMAKE = None
  CMAKE = None
  running: cd "/tmp/whisper-rs/target/debug/build/whisper-rs-sys-f7e57ece2896f8a9/out/build" && CMAKE_PREFIX_PATH="" "cmake" "-Wdev" "--debug-output" "/tmp/whisper-rs/target/debug/build/whisper-rs-sys-f7e57ece2896f8a9/out/whisper.cpp/" "-DBUILD_SHARED_LIBS=OFF" "-DWHISPER_ALL_WARNINGS=OFF" "-DWHISPER_ALL_WARNINGS_3RD_PARTY=OFF" "-DWHISPER_BUILD_TESTS=OFF" "-DWHISPER_BUILD_EXAMPLES=OFF" "-DCMAKE_INSTALL_PREFIX=/tmp/whisper-rs/target/debug/build/whisper-rs-sys-f7e57ece2896f8a9/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_C_COMPILER=/usr/bin/cc" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_CXX_COMPILER=/usr/bin/c++" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_ASM_COMPILER=/usr/bin/cc" "-DCMAKE_BUILD_TYPE=Release" "-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON"
  Running with debug output on.

  --- stderr
  wrapper.h:1:10: fatal error: 'whisper.h' file not found
  CMake Warning:
    Ignoring extra path from command line:

     "/tmp/whisper-rs/target/debug/build/whisper-rs-sys-f7e57ece2896f8a9/out/whisper.cpp/"


  CMake Error: The source directory "/tmp/whisper-rs/target/debug/build/whisper-rs-sys-f7e57ece2896f8a9/out/whisper.cpp" does not appear to contain CMakeLists.txt.
  Specify --help for usage, or press the help button on the CMake GUI.
  thread 'main' panicked at /home/niko/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cmake-0.1.50/src/lib.rs:1098:5:

  command did not execute successfully, got: exit status: 1

  build script failed, must exit now
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by npatsakula** [comment]
> Can't get this to compile locally:
> 
> ```
> memory 19GiB/62GiB at 10-15 09:01:26 exit 101 took 6s156ms
> on git sandbox_friendly is pkg v0.9.0-rc.2 via rs v1.75.0-nightly
> niko@niko-desktop whisper-rs > cargo run --example basic_use
>    Compiling whisper-rs-sys v0.7.3 (/tmp/whisper-rs/sys)
> The following warnings were emitted during compilation:
> 
> warning: Unable to generate bindings: clang diagnosed error: wrapper.h:1:10: fatal error: 'whisper.h' file not found
> warning: Using bundled bindings.rs, which may be out of date
> 
> error: failed to run custom build command for `whisper-rs-sys v0.7.3 (/tmp/whisper-rs/sys)`
> 
> Caused by:
>   process didn't exit successfully: `/tmp/whisper-rs/target/debug/build/whisper-rs-sys-c3f9ea4e385de8cf/build-script-build` (exit status: 101)
>   --- stdout
>   cargo:rustc-link-lib=dylib=stdc++
>   cargo:rerun-if-changed=wrapper.h
>   cargo:rerun-if-env-changed=TARGET
>   cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS_x86_64-unknown-linux-gnu
>   cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS_x86_64_unknown_linux_gnu
>   cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS
>   cargo:warning=Unable to generate bindings: clang diagnosed error: wrapper.h:1:10: fatal error: 'whisper.h' file not found
> 
>   cargo:warning=Using bundled bindings.rs, which may be out of date
>   CMAKE_TOOLCHAIN_FILE_x86_64-unknown-linux-gnu = None
>   CMAKE_TOOLCHAIN_FILE_x86_64_unknown_linux_gnu = None
>   HOST_CMAKE_TOOLCHAIN_FILE = None
>   CMAKE_TOOLCHAIN_FILE = None
>   CMAKE_GENERATOR_x86_64-unknown-linux-gnu = None
>   CMAKE_GENERATOR_x86_64_unknown_linux_gnu = None
>   HOST_CMAKE_GENERATOR = None
>   CMAKE_GENERATOR = None
>   CMAKE_PREFIX_PATH_x86_64-unknown-linux-gnu = None
>   CMAKE_PREFIX_PATH_x86_64_unknown_linux_gnu = None
>   HOST_CMAKE_PREFIX_PATH = None
>   CMAKE_PREFIX_PATH = None
>   CMAKE_x86_64-unknown-linux-gnu = None
>   CMAKE_x86_64_unknown_linux_gnu = None
>   HOST_CMAKE = None
>   CMAKE = None
>   running: cd "/tmp/whisper-rs/target/debug/build/whisper-rs-sys-f7e57ece2896f8a9/out/build" && CMAKE_PREFIX_PATH="" "cmake" "-Wdev" "--debug-output" "/tmp/whisper-rs/target/debug/build/whisper-rs-sys-f7e57ece2896f8a9/out/whisper.cpp/" "-DBUILD_SHARED_LIBS=OFF" "-DWHISPER_ALL_WARNINGS=OFF" "-DWHISPER_ALL_WARNINGS_3RD_PARTY=OFF" "-DWHISPER_BUILD_TESTS=OFF" "-DWHISPER_BUILD_EXAMPLES=OFF" "-DCMAKE_INSTALL_PREFIX=/tmp/whisper-rs/target/debug/build/whisper-rs-sys-f7e57ece2896f8a9/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_C_COMPILER=/usr/bin/cc" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_CXX_COMPILER=/usr/bin/c++" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_ASM_COMPILER=/usr/bin/cc" "-DCMAKE_BUILD_TYPE=Release" "-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON"
>   Running with debug output on.
> 
>   --- stderr
>   wrapper.h:1:10: fatal error: 'whisper.h' file not found
>   CMake Warning:
>     Ignoring extra path from command line:
> 
>      "/tmp/whisper-rs/target/debug/build/whisper-rs-sys-f7e57ece2896f8a9/out/whisper.cpp/"
> 
> 
>   CMake Error: The source directory "/tmp/whisper-rs/target/debug/build/whisper-rs-sys-f7e57ece2896f8a9/out/whisper.cpp" does not appear to contain CMakeLists.txt.
>   Specify --help for usage, or press the help button on the CMake GUI.
>   thread 'main' panicked at /home/niko/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cmake-0.1.50/src/lib.rs:1098:5:
> 
>   command did not execute successfully, got: exit status: 1
> 
>   build script failed, must exit now
>   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```

Have you downloaded submodules?

```bash
git submodule update --init --recursive
```

I tested this setup in few environments: OS X M1, X86 and ARM servers.

**Comment by tazz4843** [comment]
I always forget to clone the submodules, thanks for reminding me. Seems to work now.=

**Comment by tazz4843** [comment]
Thanks for the PR!

**Pull request merged**